internal/tracker/upstream.go

```diff
@@ -0,0 +1,196 @@
+package tracker
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+var errorNoAvailableUpstream = errors.New("No Available Upstream")
+
+// UpstreamConns tracks connections for an upstreamGroup
+// on a per upstream basis. Upstreams can be marked as
+// unhealthy to prevent them from being chosen for new connections.
+// UpstreamConns handles load balancing through BeginConnection()
+type UpstreamConns struct {
+	// mu protects the resources of UpstreamConns
+	mu sync.Mutex
+
+	// upstreams holds all upstreams, healthy or unhealthy
+	upstreams map[uuid.UUID]*upstream
+
+	// pq holds healthy upstreams and provides the means to
+	// pick the upstream with the least connections.
+	pq *upstreamPQ
+}
+
+// An upstream stores a count of connections
+// as well as some overhead for use in upstreamPQ.
+type upstream struct {
+	// id is the id of the upstream
+	id uuid.UUID
+
+	// The count of connections to the upstream.
+	// Also the priority of an upstream, lowest first.
+	connCount uint32
+
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	// if an upstream is pulled from the upstreamPQ (because of health)
+	// its index will be set to -1
+	index int
+}
+
+// NewUpstreamConns creates a new UpstreamConns
+// with upstreams based on provided upstreamIDs.
+// upstreams must be marked as healthy before they will be
+// added to the internal priorityQueue and available for BeginConnection()
+func NewUpstreamConns(upstreamIDs []uuid.UUID) *UpstreamConns {
+	upstreams := map[uuid.UUID]*upstream{}
+	for _, id := range upstreamIDs {
+		upstreams[id] = &upstream{
+			id:    id,
+			index: -1,
+		}
+	}
+	return &UpstreamConns{
+		upstreams: upstreams,
+		pq:        &upstreamPQ{},
+	}
+}
+
+// BeginConnection returns the UUID of the upstream with the least connections
+// and records the additional connection.
+// An error is returned if there are no available upstreams
+func (t *UpstreamConns) BeginConnection() (uuid.UUID, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	upstream := t.pq.peek()
+	if upstream == nil {
+		return uuid.UUID{}, errorNoAvailableUpstream
+	}
+
+	// do we need a check for an upstream which is not in the upstreamPQ?
+	// The assumption is that we are only incrementing upstreams which are
+	// healthy and in the upstreamPQ. unhealthy upstreams are removed from the upstreamPQ.
+	upstream.connCount += 1
+	heap.Fix(t.pq, upstream.index)
+	return upstream.id, nil
+}
+
+// EndConnection takes the UUID of the upstream which has just had a connection terminate
+// and records the ended connection.
+func (t *UpstreamConns) EndConnection(id uuid.UUID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	upstream := t.upstreams[id]
```

Comment:
A nil pointer dereference.

Thanks for catching this. I left a comment on downstreamConns.EndConnection(), but forgot it here. This one has worse consequences too.

I leaned towards specifying the contract because the package is internal and the code will likely be executed more often. Would you rather see a check?

Reply:
I think reasons not to are sparing. If I had to point out a place in this package that would have the best argument for depending on contract, it might be this. Its also not justification enough. Thanks for catching it